### DISCUSSION TITLE
What does "-d" debug flag have on the UDP test?

### DISCUSSION THREAD

**Discussion by PraptiSarker02** [esnet/iperf#1424] [Q&A] [answered]
I am using iperf3.12 on windows. Willing to know what does -d flag do in background and why does it reduce the bitrate and lost?

iperf command used: iperf3 -c 192.168.1.93 -t 10 -w4M -u -b 1G -d -V
![UDP_with debug](https://user-images.githubusercontent.com/64602885/203123516-0aa94f60-caac-4769-83ad-599334e6523b.png)

iperf command used: iperf3 -c 192.168.1.93 -t 10 -w4M -u -b 1G -V
![UDP_without_debug](https://user-images.githubusercontent.com/64602885/203123790-e5fad264-1622-48ca-afd8-a3018cd9f45a.png)

**Comment by davidBar-On** [comment, marked as answer]
`-d` prints a line for each message sent, and is equivalent to `--debug` or `--debug=4`.  Printing all these message is a big overhead.  If you don't need the printed line for each message sent (which is usually the case) **use `--debug=3`**.  In this debug level, the lines per message are not printed and hopefully this will resolve the performance issue.

**Reply by PraptiSarker02** [reply]
Yes, "--debug=3" is giving me the same 50% drop. Please help me understand the "big overhead" term. So, you meant, due to high volume of printing, the bit rate has reduced, and as a result the network was not congested and there wasn't any drop packets.

**Comment by davidBar-On** [comment]
> ... due to high volume of printing, the bit rate has reduced, and as a result the network was not congested and there wasn't any drop packets.

Correct.  The bitrate was reduced because the client sending rate was reduced.  With the `--debug=3` only few debug messages are printed per second, so it did not affect the sending rate.  With UDP, only the `-b` and the CPU performance affect the sending rate.

The only thing that I don't understand is the 49% lost rate which is almost exactly 50% and seems suspicious.  Can you try running a TCP test without rate limit (no `-b`) and with `-l 1460` (same packet length as used for UDP)?  The result throughput will be close the maximum network throughput rate, which should be close to the 377Mbits/sec (the UDP throughput without the `-d`) if the debug messages are the main reason for the lost packets.

**Reply by PraptiSarker02** [reply]
Thanks @davidBar-On for getting back. The tcp test throughput is around 950Mbps. I would also request you to look at discussion #1422, it is something the "-l" length of packet/datagram, increasing the packet length to 8000byte for udp test reduce the drop and also increases the throughput close to what I get from tcp test. I will very thankful if you also look at the discussion #1422.

**Reply by PraptiSarker02** [reply]
Actually, my bad. tcp test with "-l 1460" is around 650Mbps but without is 900Mbps.

iperf command used: iperf3 -c 192.168.0.2 -t 10 -w4M -l 1460
![image](https://user-images.githubusercontent.com/64602885/203153238-8467af2e-613c-48d6-a141-8b2653825908.png)

iperf command used: iperf3 -c 192.168.0.2 -t 10 -w4M
![image](https://user-images.githubusercontent.com/64602885/203153368-507ba8f2-d168-44a7-8798-97c0e441bba7.png)

**Reply by PraptiSarker02** [reply]
As you can see the ip address has changes, as previously there was a router in between two PC, but now two PC are connected directly by ethernet.

**Reply by PraptiSarker02** [reply]
The TCP result actually got me thinking shouldn't in both cases: "-l 1460" and default 128KB, gives same throughput. The low throughput with "-l 1460" makes no sense if the tcp protocol is dividing that 128KB in chunks of 1460 bytes(as I talked in discussion #1422). So what what is going on?

From the result the only explanation I get is: at "-l 1460" TCP is creating more packets and due to high volume of packet, network get congested, resulting resend of some packet and causing the throughput to reduce. But with that explanation isn't 128KB divided into 1460 bytes of payload? The MTU is 1500 byte. How does everything works @davidBar-On?

**Comment by davidBar-On** [comment]
The reduced TCP throughput with smaller packets is because of the iperf3 and the system has an overhead for sending each packet - calling the send function, copying the packets to the system buffers, etc.  With 1460 bytes packets, iperf3 sends about 90 times more packets than with 128KB packets for the same throughput.  AS @bmah888 explained in #1422, once the packet data is in the system TCP buffers, there is no meaning to the packet size.

Using `--zerocopy` (works only for TCP) should help with small packets.  `-b /30` may also help - "30" is the the burst rate for sending packets - 10 is the default (works for both TCP and UDP).

**Reply by PraptiSarker02** [reply]
@davidBar-On, Please let me know if I am think right. So, in case of 128KB the system buffer filling command is called less frequently than when the block size is set to 1460byte. Which means the TCP buffer is filled more quickly in case of 128KB compare to 1460B, due to overhead and also because of the filling length(128KB vs 1260B). The question arise here is: in both cases how the data is transmitted? I thought in both cases data is transmitted a chunk of 1500bytes[which contradict with your statement: "With 1460 bytes packets, iperf3 sends about 90 times more packets than with 128KB packets for the same throughput"]. I am getting the fact that in case the of 1460B there is some overhead delay  but shouldn't the transmission speed be same? So, this throughput drop around 300Mbps is due that system overhead?

**Reply by davidBar-On** [reply]
> ... in case the of 1460B there is some overhead delay but shouldn't the transmission speed be same? So, this throughput drop around 300Mbps is due that system overhead?

The overhead is causing the UDP send rate to reduce from about 950mbps for 8000B packets to about 750mbps (and TCP form 950MB for 128KB packets to 670mbps for 1460B packets).  It is probably not related to network performance.  See below for more details.

> So, in case of 128KB the system buffer filling command is called less frequently than when the block size is set to 1460byte. Which means the TCP buffer is filled more quickly in case of 128KB compare to 1460B, due to overhead and also because of the filling length(128KB vs 1260B).

In general, you are right about the "buffer filling", but the overhead is also (or mainly) related to other activities.  Following is the general iperf3 process for sending a "burst" (by default 10) of packets wich probably causes most of the overhead:
1. [Call system `select()`](https://github.com/esnet/iperf/blob/a2ce47a97acfad3acee91a6cb4da28f9f50c19df/src/iperf_client_api.c#L587) function to see if any input arrived.
2. [Call `iperf_send()`](https://github.com/esnet/iperf/blob/a2ce47a97acfad3acee91a6cb4da28f9f50c19df/src/iperf_client_api.c#L640) to send the next burst of packets.
3. In [`iperf_send()`](https://github.com/esnet/iperf/blob/a2ce47a97acfad3acee91a6cb4da28f9f50c19df/src/iperf_api.c#L1860-L1918) burst loop, [call the send function](https://github.com/esnet/iperf/blob/a2ce47a97acfad3acee91a6cb4da28f9f50c19df/src/iperf_api.c#L1889) (UDP/TCP/STCP) for each packet.
5. UDP send function (as an example) [calls `Nwrite()`](https://github.com/esnet/iperf/blob/a2ce47a97acfad3acee91a6cb4da28f9f50c19df/src/iperf_udp.c#L251) to send the packet.
6. `Nwrite()` calls the system call `write()` to send the packet.

In addition, for each packet statistics are collected, a check is done if exceeded bitrate, etc.
